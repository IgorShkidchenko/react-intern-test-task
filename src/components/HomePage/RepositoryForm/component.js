import React from "react";
import PropTypes from "prop-types";
import { Form, Field, ErrorMessage } from "formik";

const RepositoryFormComponent = ({ isSubmitting }) => (
  <Form>
    <Field type="repositoryUrl" name="repositoryUrl" placeholder="Enter repository name e.g. facebook/react" />
    <ErrorMessage name="repositoryUrl" component="p" />
    <button type="submit" disabled={isSubmitting}>
      Submit
    </button>
  </Form>
);

RepositoryFormComponent.propTypes = {
  isSubmitting: PropTypes.bool.isRequired,
};

export default RepositoryFormComponent;
