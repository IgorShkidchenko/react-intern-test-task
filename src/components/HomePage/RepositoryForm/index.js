import React from "react";
import PropTypes from "prop-types";
import axios from "axios";
import { Formik, Form, Field, ErrorMessage } from "formik";

const RepositoryForm = ({ onAddRepository }) => (
  <Formik
    initialValues={{ repository: "" }}
    validate={(values) => {
      const errors = {};
      if (!values.repository) {
        errors.repository = "Required";
      } else if (
        !/^\w+\/{1}\w+$/.test(values.repository)
      ) {
        errors.repository = "Invalid repository address";
      }
      return errors;
    }}
    onSubmit={(values, { setSubmitting, setErrors }) => {
      const url = `https://api.github.com/repos/${values.repository}`;
      let promiseResponse;

      axios.get(url)
        .then((response) => {
          promiseResponse = response.data;
          onAddRepository(promiseResponse);
        })
        .catch(() => {
          setErrors({ repository: "Repository not found" });
        });
      setSubmitting(false);
    }}
  >
    {({ isSubmitting }) => (
      <Form>
        <Field type="repository" name="repository" placeholder="Enter repository name e.g. facebook/react" />
        <ErrorMessage name="repository" component="div" />
        <button type="submit" disabled={isSubmitting}>
          Submit
        </button>
      </Form>
    )}
  </Formik>
);

RepositoryForm.propTypes = {
  onAddRepository: PropTypes.func.isRequired,
};

export default RepositoryForm;
