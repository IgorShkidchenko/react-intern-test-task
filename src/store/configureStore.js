import axios from "axios";
import rootReducer from "./rootReducer";
import logic from "./rootLogic";

import { createStore, applyMiddleware } from "redux";
import { createLogicMiddleware } from "redux-logic";

const deps = { httpClient: axios };
const logicMiddleware = createLogicMiddleware(logic, deps);

const configureStore = () => createStore(rootReducer, applyMiddleware(logicMiddleware));

export default configureStore;
