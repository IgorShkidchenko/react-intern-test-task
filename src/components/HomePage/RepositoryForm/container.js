import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import { Formik, Form, Field, ErrorMessage } from "formik";

import { repositoryActions } from "../../../store/repositories";

export class RepositoryForm extends React.Component {
  render() {
    return (
      <Formik
        initialValues={{ repositoryUrl: "" }}
        validate={(values) => {
          const errors = {};
          if (!values.repositoryUrl) {
            errors.repositoryUrl = "Required";
          } else if (
            !/^.+\/{1}.+$/.test(values.repositoryUrl)
          ) {
            errors.repositoryUrl = "Invalid repository address";
          }
          return errors;
        }}
        onSubmit={(values, { setSubmitting, setErrors }) => {
          const payload = {
            repositoryUrl: values.repositoryUrl,
            setSubmitting,
            setErrors,
          };
          this.props.onAddRepository(payload);
        }}
      >
        {({ isSubmitting }) => (
          <Form>
            <Field type="repositoryUrl" name="repositoryUrl" placeholder="Enter repository name e.g. facebook/react" />
            <ErrorMessage name="repositoryUrl" component="p" />
            <button type="submit" disabled={isSubmitting}>
              Submit
            </button>
          </Form>
        )}
      </Formik>
    );
  }
}

RepositoryForm.propTypes = {
  onAddRepository: PropTypes.func.isRequired,
};

const mapDispatchToProps = dispatch => ({
  onAddRepository: payload => dispatch(repositoryActions.getRepository(payload)),
});

export default connect(null, mapDispatchToProps)(RepositoryForm);
