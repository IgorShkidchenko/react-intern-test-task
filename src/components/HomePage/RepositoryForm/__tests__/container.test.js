import React from "react";
import { shallow } from "enzyme";
import { waitFor } from "@testing-library/dom";
import configureStore from "redux-mock-store";

import RepositoryForm from "../container";

describe("<RepotisoryRow />", () => {
  const clickFn = jest.fn();
  const mockStore = configureStore([]);
  const initialState = { onAddRepository: clickFn };
  const store = mockStore(initialState);

  let container;
  beforeEach(() => {
    const wrapper = shallow(<RepositoryForm store={store} />);
    container = wrapper.dive().dive().dive();
  });

  it("renders correctly", () => {
    expect(container).toMatchSnapshot();
  });

  it("should call onAddRepository on submit", async () => {
    const value = { target: { name: "repositoryUrl", value: "facebook/react" } };
    container.find("input").simulate("change", value);

    await waitFor(() => {
      container.find("form").simulate("submit");
      expect(clickFn).toHaveBeenCalledTimes(1);
    });
  });
});
