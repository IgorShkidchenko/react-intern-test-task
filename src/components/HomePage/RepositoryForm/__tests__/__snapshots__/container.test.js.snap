// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<RepotisoryRow /> renders correctly 1`] = `
<Form>
  <Field
    name="repositoryUrl"
    placeholder="Enter repository name e.g. facebook/react"
    type="repositoryUrl"
  />
  <FormikConnect(ErrorMessageImpl)
    component="p"
    name="repositoryUrl"
  />
  <button
    disabled={false}
    type="submit"
  >
    Submit
  </button>
</Form>
`;
